File: AI-agent-backend/app/entity/api_endpoint.py
# Copyright (c) 2025 左岚. All rights reserved.
"""
API端点实体
定义API接口管理的数据模型
"""

from datetime import datetime
from typing import Optional, Dict, Any
from enum import Enum

from sqlalchemy import Column, Integer, String, Text, DateTime, ForeignKey, JSON, Float
from sqlalchemy.orm import relationship

from .base import BaseEntity


class ApiMethod(str, Enum):
    """API请求方法枚举"""
    GET = "GET"
    POST = "POST"
    PUT = "PUT"
    DELETE = "DELETE"
    PATCH = "PATCH"
    HEAD = "HEAD"
    OPTIONS = "OPTIONS"


class ApiStatus(str, Enum):
    """API状态枚举"""
    ACTIVE = "active"  # 激活
    INACTIVE = "inactive"  # 未激活
    DEPRECATED = "deprecated"  # 已废弃
    MAINTENANCE = "maintenance"  # 维护中


class ApiEndpoint(BaseEntity):
    """
    API端点实体类
    定义API接口的基本信息和统计数据
    """
    __tablename__ = "api_endpoint"
    __allow_unmapped__ = True  # 允许未映射的注解

    # API路径 - 必填，最大500个字符
    path = Column(String(500), nullable=False, comment="API路径")

    # API方法 - 必填，使用枚举值
    method = Column(String(10), nullable=False, comment="HTTP方法")

    # API名称 - 必填，最大100个字符
    name = Column(String(100), nullable=False, comment="API名称")

    # API描述 - 可选，文本类型
    description = Column(Text, nullable=True, comment="API描述")

    # API状态 - 必填，使用枚举值
    status = Column(String(20), nullable=False, default=ApiStatus.ACTIVE.value, comment="API状态")

    # 所属模块 - 可选，最大100个字符
    module = Column(String(100), nullable=True, comment="所属模块")

    # 权限标识 - 可选，最大100个字符
    permission = Column(String(100), nullable=True, comment="权限标识")

    # API版本 - 可选，最大20个字符
    version = Column(String(20), nullable=True, default="v1", comment="API版本")

    # 请求参数示例 - JSON格式存储
    request_example = Column(JSON, nullable=True, comment="请求参数示例")

    # 响应示例 - JSON格式存储
    response_example = Column(JSON, nullable=True, comment="响应示例")

    # 访问统计
    total_calls = Column(Integer, default=0, comment="总调用次数")
    success_calls = Column(Integer, default=0, comment="成功调用次数")
    error_calls = Column(Integer, default=0, comment="错误调用次数")

    # 性能统计
    avg_response_time = Column(Float, default=0.0, comment="平均响应时间(ms)")
    max_response_time = Column(Float, default=0.0, comment="最大响应时间(ms)")
    min_response_time = Column(Float, default=0.0, comment="最小响应时间(ms)")

    # 最后调用时间
    last_called_at = Column(DateTime, nullable=True, comment="最后调用时间")

    # 创建者ID - 外键关联用户表
    created_by_id = Column(Integer, ForeignKey('user.id'), nullable=False, comment="创建者ID")

    # 关联关系
    # API-创建者关联
    creator = relationship("User", foreign_keys=[created_by_id])

    def __init__(self, path: str, method: str, name: str, description: str = None,
                 module: str = None, permission: str = None, version: str = "v1",
                 request_example: Dict[str, Any] = None, response_example: Dict[str, Any] = None,
                 created_by_id: int = None):
        """
        初始化API端点

        Args:
            path: API路径
            method: HTTP方法
            name: API名称
            description: API描述
            module: 所属模块
            permission: 权限标识
            version: API版本
            request_example: 请求参数示例
            response_example: 响应示例
            created_by_id: 创建者ID
        """
        self.path = path
        self.method = method
        self.name = name
        self.description = description
        self.status = ApiStatus.ACTIVE.value
        self.module = module
        self.permission = permission
        self.version = version
        self.request_example = request_example or {}
        self.response_example = response_example or {}
        self.created_by_id = created_by_id
        self.total_calls = 0
        self.success_calls = 0
        self.error_calls = 0
        self.avg_response_time = 0.0
        self.max_response_time = 0.0
        self.min_response_time = 0.0

    def is_active(self) -> bool:
        """判断API是否激活"""
        return self.status == ApiStatus.ACTIVE.value

    def is_deprecated(self) -> bool:
        """判断API是否已废弃"""
        return self.status == ApiStatus.DEPRECATED.value

    def get_success_rate(self) -> float:
        """获取成功率"""
        if self.total_calls == 0:
            return 0.0
        return (self.success_calls / self.total_calls) * 100

    def update_stats(self, success: bool, response_time: float):
        """更新统计数据"""
        self.total_calls += 1
        self.last_called_at = datetime.utcnow()
        
        if success:
            self.success_calls += 1
        else:
            self.error_calls += 1
        
        # 更新响应时间统计
        if self.total_calls == 1:
            self.avg_response_time = response_time
            self.max_response_time = response_time
            self.min_response_time = response_time
        else:
            # 计算新的平均响应时间
            total_time = self.avg_response_time * (self.total_calls - 1) + response_time
            self.avg_response_time = total_time / self.total_calls
            
            # 更新最大最小响应时间
            if response_time > self.max_response_time:
                self.max_response_time = response_time
            if response_time < self.min_response_time:
                self.min_response_time = response_time

    def to_dict(self) -> Dict[str, Any]:
        """转换为字典格式"""
        return {
            "id": self.id,
            "path": self.path,
            "method": self.method,
            "name": self.name,
            "description": self.description,
            "status": self.status,
            "module": self.module,
            "permission": self.permission,
            "version": self.version,
            "request_example": self.request_example,
            "response_example": self.response_example,
            "total_calls": self.total_calls,
            "success_calls": self.success_calls,
            "error_calls": self.error_calls,
            "success_rate": self.get_success_rate(),
            "avg_response_time": self.avg_response_time,
            "max_response_time": self.max_response_time,
            "min_response_time": self.min_response_time,
            "last_called_at": self.last_called_at.isoformat() if self.last_called_at else None,
            "created_at": self.created_at.isoformat() if self.created_at else None,
            "updated_at": self.updated_at.isoformat() if self.updated_at else None,
            "created_by_id": self.created_by_id
        }

File: AI-agent-frontend/src/api/modules/apiEndpoint.ts
/**
 * API端点管理接口
 */

import http from '@/api/http'

// API端点相关接口类型定义
export interface ApiEndpoint {
  id: number
  path: string
  method: string
  name: string
  description?: string
  status: string
  module?: string
  permission?: string
  version: string
  request_example?: Record<string, any>
  response_example?: Record<string, any>
  total_calls: number
  success_calls: number
  error_calls: number
  success_rate: number
  avg_response_time: number
  max_response_time: number
  min_response_time: number
  last_called_at?: string
  created_at?: string
  updated_at?: string
  created_by_id: number
}

export interface ApiEndpointCreateRequest {
  path: string
  method: string
  name: string
  description?: string
  module?: string
  permission?: string
  version?: string
  request_example?: Record<string, any>
  response_example?: Record<string, any>
}

export interface ApiEndpointUpdateRequest {
  name?: string
  description?: string
  status?: string
  module?: string
  permission?: string
  version?: string
  request_example?: Record<string, any>
  response_example?: Record<string, any>
}

export interface ApiEndpointQueryParams {
  page?: number
  size?: number
  keyword?: string
  method?: string
  status?: string
  module?: string
  permission?: string
}

export interface ApiEndpointListResponse {
  items: ApiEndpoint[]
  total: number
  page: number
  size: number
  pages: number
}

export interface ApiStatistics {
  total_apis: number
  active_apis: number
  deprecated_apis: number
  maintenance_apis: number
  total_calls_today: number
  success_calls_today: number
  error_calls_today: number
  avg_response_time: number
  top_apis: Array<{
    name: string
    path: string
    calls: number
  }>
  error_apis: Array<{
    name: string
    path: string
    error_rate: number
  }>
}

export interface ApiCallLog {
  id: number
  api_id: number
  api_path: string
  method: string
  status_code: number
  response_time: number
  user_id?: number
  ip_address: string
  user_agent?: string
  request_data?: Record<string, any>
  response_data?: Record<string, any>
  error_message?: string
  created_at: string
}

/**
 * API端点管理API类
 */
export class ApiEndpointApi {
  /**
   * 创建API端点
   */
  static async createApiEndpoint(data: ApiEndpointCreateRequest) {
    return http.post<ApiEndpoint>('/api/v1/api-endpoints/', data)
  }

  /**
   * 更新API端点
   */
  static async updateApiEndpoint(id: number, data: ApiEndpointUpdateRequest) {
    return http.put<ApiEndpoint>(`/api/v1/api-endpoints/${id}`, data)
  }

  /**
   * 删除API端点
   */
  static async deleteApiEndpoint(id: number) {
    return http.delete<boolean>(`/api/v1/api-endpoints/${id}`)
  }

  /**
   * 获取API端点详情
   */
  static async getApiEndpoint(id: number) {
    return http.get<ApiEndpoint>(`/api/v1/api-endpoints/${id}`)
  }

  /**
   * 获取API端点列表
   */
  static async getApiEndpoints(params?: ApiEndpointQueryParams) {
    return http.get<ApiEndpointListResponse>('/api/v1/api-endpoints/', { params })
  }

  /**
   * 获取API统计数据
   */
  static async getApiStatistics() {
    return http.get<ApiStatistics>('/api/v1/api-endpoints/statistics/overview')
  }

  /**
   * 获取所有模块列表
   */
  static async getModules() {
    return http.get<string[]>('/api/v1/api-endpoints/metadata/modules')
  }

  /**
   * 获取所有权限列表
   */
  static async getPermissions() {
    return http.get<string[]>('/api/v1/api-endpoints/metadata/permissions')
  }

  /**
   * 获取所有HTTP方法列表
   */
  static async getMethods() {
    return http.get<string[]>('/api/v1/api-endpoints/metadata/methods')
  }

  /**
   * 批量更新API状态
   */
  static async batchUpdateStatus(apiIds: number[], status: string) {
    return http.post<number>('/api/v1/api-endpoints/batch/status', {
      api_ids: apiIds,
      status
    })
  }

  /**
   * 从路由同步API端点
   */
  static async syncApiEndpoints() {
    return http.post<{
      total_scanned: number
      new_created: number
      updated: number
      skipped: number
    }>('/api/v1/api-endpoints/sync')
  }

  /**
   * 获取API调用日志
   */
  static async getApiCallLogs(params?: {
    page?: number
    size?: number
    api_id?: number
    start_date?: string
    end_date?: string
  }) {
    return http.get<{
      items: ApiCallLog[]
      total: number
      page: number
      size: number
      pages: number
    }>('/api/v1/api-endpoints/logs', { params })
  }
}

// 导出单个方法，保持兼容性
export const apiEndpointApi = {
  createApiEndpoint: ApiEndpointApi.createApiEndpoint,
  updateApiEndpoint: ApiEndpointApi.updateApiEndpoint,
  deleteApiEndpoint: ApiEndpointApi.deleteApiEndpoint,
  getApiEndpoint: ApiEndpointApi.getApiEndpoint,
  getApiEndpoints: ApiEndpointApi.getApiEndpoints,
  getApiStatistics: ApiEndpointApi.getApiStatistics,
  getModules: ApiEndpointApi.getModules,
  getPermissions: ApiEndpointApi.getPermissions,
  getMethods: ApiEndpointApi.getMethods,
  batchUpdateStatus: ApiEndpointApi.batchUpdateStatus,
  syncApiEndpoints: ApiEndpointApi.syncApiEndpoints,
  getApiCallLogs: ApiEndpointApi.getApiCallLogs
}

// 默认导出
export default ApiEndpointApi

// API状态常量
export const API_STATUS = {
  ACTIVE: 'active',
  INACTIVE: 'inactive',
  DEPRECATED: 'deprecated',
  MAINTENANCE: 'maintenance'
} as const

// API状态标签映射
export const API_STATUS_LABELS = {
  [API_STATUS.ACTIVE]: '激活',
  [API_STATUS.INACTIVE]: '未激活',
  [API_STATUS.DEPRECATED]: '已废弃',
  [API_STATUS.MAINTENANCE]: '维护中'
} as const

// HTTP方法常量
export const HTTP_METHODS = {
  GET: 'GET',
  POST: 'POST',
  PUT: 'PUT',
  DELETE: 'DELETE',
  PATCH: 'PATCH',
  HEAD: 'HEAD',
  OPTIONS: 'OPTIONS'
} as const

// HTTP方法颜色映射
export const HTTP_METHOD_COLORS = {
  [HTTP_METHODS.GET]: 'success',
  [HTTP_METHODS.POST]: 'primary',
  [HTTP_METHODS.PUT]: 'warning',
  [HTTP_METHODS.DELETE]: 'danger',
  [HTTP_METHODS.PATCH]: 'info',
  [HTTP_METHODS.HEAD]: 'secondary',
  [HTTP_METHODS.OPTIONS]: 'secondary'
} as const
